// @angular-mdl/core/esm2015/lib/progress/mdl-progress.component.js
import { ChangeDetectionStrategy, Component, HostBinding, Input, ViewEncapsulation, } from "@angular/core";
import { toBoolean } from "../common/boolean-property";
export class MdlProgressComponent {
    constructor() {
        this.progress = 0;
        this.buffer = 100;
        this.aux = 0;
        this.isProgess = true;
        this.indeterminateIntern = false;
    }
    get indeterminate() {
        return this.indeterminateIntern;
    }
    set indeterminate(value) {
        this.indeterminateIntern = toBoolean(value);
    }
    ngOnChanges(changes) {
        if (changes.buffer) {
            this.setBuffer(changes.buffer.currentValue);
        }
    }
    setBuffer(b) {
        this.aux = 100 - b;
    }
}
MdlProgressComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-progress",
                template: `
    <div class="progressbar bar bar1" [style.width]="progress + '%'"></div>
    <div class="bufferbar bar bar2" [style.width]="buffer + '%'"></div>
    <div class="auxbar bar bar3" [ngStyle]="{ width: aux + '%' }"></div>
  `,
                encapsulation: ViewEncapsulation.None,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
MdlProgressComponent.propDecorators = {
    progress: [{ type: Input }],
    buffer: [{ type: Input }],
    aux: [{ type: Input }],
    isProgess: [{ type: HostBinding, args: ["class.mdl-progress",] }],
    indeterminate: [{ type: HostBinding, args: ["class.mdl-progress__indeterminate",] }, { type: Input }]
};
