// @angular-mdl/core/esm2015/lib/dialog/mdl-simple-dialog.component.js
import { Component, forwardRef, HostListener, Inject, QueryList, ViewChildren, ViewEncapsulation, } from "@angular/core";
import { MDL_CONFIGUARTION } from "./config";
import { MdlButtonComponent } from "../button/mdl-button.component";
import { MdlDialogReference } from "./mdl-dialog-reference";
export class MdlSimpleDialogComponent {
    // why do i need forwardRef at this point, the demo LoginDialog dosn't need this!?!?
    constructor(dialogConfiguration, dialog) {
        this.dialogConfiguration = dialogConfiguration;
        this.dialog = dialog;
        this.dialog = dialog;
        dialog.onVisible().subscribe(() => {
            if (this.buttons) {
                this.buttons.first.elementRef.nativeElement.focus();
            }
        });
    }
    onEsc() {
        // run the first action that is marked as closing action
        const closeAction = this.dialogConfiguration.actions.find((action) => action.isClosingAction);
        if (closeAction) {
            closeAction.handler();
            this.dialog.hide();
        }
    }
    actionClicked(action) {
        action.handler();
        this.dialog.hide();
    }
}
MdlSimpleDialogComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-dialog-component",
                template: `
    <h3 class="mdl-dialog__title" *ngIf="dialogConfiguration?.title">
      {{ dialogConfiguration?.title }}
    </h3>
    <div
      class="mdl-dialog__content"
      [innerHTML]="dialogConfiguration?.message"
    ></div>
    <div
      class="mdl-dialog__actions"
      [ngClass]="{
        'mdl-dialog__actions--full-width': dialogConfiguration?.fullWidthAction
      }"
    >
      <button
        mdl-button
        mdl-colored="primary"
        type="button"
        *ngFor="let action of dialogConfiguration?.actions"
        (click)="actionClicked(action)"
        [ngClass]="{ close: action.isClosingAction }"
      >
        {{ action.text }}
      </button>
    </div>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlSimpleDialogComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [forwardRef(() => MDL_CONFIGUARTION),] }] },
    { type: MdlDialogReference, decorators: [{ type: Inject, args: [forwardRef(() => MdlDialogReference),] }] }
];
MdlSimpleDialogComponent.propDecorators = {
    buttons: [{ type: ViewChildren, args: [MdlButtonComponent,] }],
    onEsc: [{ type: HostListener, args: ["keydown.esc",] }]
};
