// @angular-mdl/core/esm2015/lib/dialog/mdl-dialog.component.js
import { Component, EventEmitter, Input, Output, TemplateRef, ViewChild, ViewEncapsulation, } from "@angular/core";
import { Subject } from "rxjs";
import { MdlDialogService } from "./mdl-dialog.service";
export class MdlDialogComponent {
    constructor(dialogService) {
        this.dialogService = dialogService;
        // eslint-disable-next-line
        this.showEmitter = new EventEmitter();
        // eslint-disable-next-line
        this.hideEmitter = new EventEmitter();
        this.isShown = false;
        this.dialogRef = null;
    }
    show() {
        if (this.isShown) {
            throw new Error("Only one instance of an embedded mdl-dialog can exist!");
        }
        this.isShown = true;
        const mergedConfig = this.config || {};
        // default is true
        if (typeof mergedConfig.isModal === "undefined") {
            mergedConfig.isModal = true;
        }
        const result = new Subject();
        const p = this.dialogService.showDialogTemplate(this.template, mergedConfig);
        p.subscribe((dialogRef) => {
            this.dialogRef = dialogRef;
            this.dialogRef.onVisible().subscribe(() => {
                this.showEmitter.emit(dialogRef);
                result.next(dialogRef);
                result.complete();
            });
            this.dialogRef.onHide().subscribe(() => {
                this.hideEmitter.emit(null);
                this.dialogRef = null;
                this.isShown = false;
            });
        });
        return result.asObservable();
    }
    close() {
        if (this.dialogRef) {
            this.dialogRef.hide();
        }
    }
}
MdlDialogComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-dialog",
                template: `
    <div *dialogTemplate>
      <ng-content></ng-content>
    </div>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlDialogComponent.ctorParameters = () => [
    { type: MdlDialogService }
];
MdlDialogComponent.propDecorators = {
    template: [{ type: ViewChild, args: [TemplateRef, { static: true },] }],
    config: [{ type: Input, args: ['mdl-dialog-config',] }],
    showEmitter: [{ type: Output, args: ['show',] }],
    hideEmitter: [{ type: Output, args: ['hide',] }]
};
