// @angular-mdl/core/esm2015/lib/dialog/mdl-alert.component.js
import { Component, EventEmitter, HostBinding, Input, Output, } from "@angular/core";
import { MdlDialogService } from "./mdl-dialog.service";
export class MdlAlertComponent {
    constructor(mdlDialogService) {
        this.mdlDialogService = mdlDialogService;
        this.display = "none";
        this.confirmed = new EventEmitter();
    }
    show() {
        this.mdlDialogService
            .alert(this.message, this.okText, this.title)
            .subscribe(() => {
            this.confirmed.emit();
        });
    }
}
MdlAlertComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-alert",
                template: ``,
                exportAs: "mdlAlert"
            },] }
];
MdlAlertComponent.ctorParameters = () => [
    { type: MdlDialogService }
];
MdlAlertComponent.propDecorators = {
    title: [{ type: Input }],
    message: [{ type: Input }],
    okText: [{ type: Input }],
    display: [{ type: HostBinding, args: ["style.display",] }],
    confirmed: [{ type: Output }]
};
