// @angular-mdl/core/esm2015/lib/menu/mdl-menu-item.component.js
import { Component, ElementRef, forwardRef, HostBinding, HostListener, Inject, Input, ViewEncapsulation, } from "@angular/core";
import { MdlMenuComponent } from "./mdl-menu.component";
import { toBoolean } from "../common/boolean-property";
import { callNative } from "../common/native-support";
export class MdlMenuItemComponent {
    // forwardRef is needed because of he circular dependency menu queries menuitems; menuitem needs the parent
    constructor(elementRef, mdlMenu) {
        this.elementRef = elementRef;
        this.mdlMenu = mdlMenu;
        this.tabindex = -1;
        this.isMenuItem = true;
        this.disabledIntern = false;
        this.element = elementRef.nativeElement;
    }
    get disabled() {
        return this.disabledIntern;
    }
    set disabled(value) {
        this.disabledIntern = toBoolean(value);
    }
    onClick($event) {
        $event.stopPropagation();
        if (this.disabled) {
            this.mdlMenu.hide();
            return;
        }
        this.mdlMenu.hideOnItemClicked();
    }
    // we need to register a touchstart at the window to get informed if the user taps outside the menu.
    // But if we register a touchstart event - safari will no longer convert touch events to click events.
    // So we need to convert touch to click and the user still needs to register a (click) listener to be
    // informed if the menu item has clicked.
    onTouch($event) {
        // ensure that this event is totally consumed
        $event.stopPropagation();
        $event.preventDefault();
        const event = new MouseEvent("click", { bubbles: true });
        callNative(this.element, "dispatchEvent", event);
    }
}
MdlMenuItemComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-menu-item",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlMenuItemComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: MdlMenuComponent, decorators: [{ type: Inject, args: [forwardRef(() => MdlMenuComponent),] }] }
];
MdlMenuItemComponent.propDecorators = {
    tabindex: [{ type: HostBinding, args: ["tabindex",] }],
    isMenuItem: [{ type: HostBinding, args: ["class.mdl-menu__item",] }],
    disabled: [{ type: Input }],
    onClick: [{ type: HostListener, args: ["click", ["$event"],] }],
    onTouch: [{ type: HostListener, args: ["touchstart", ["$event"],] }]
};
