// @angular-mdl/core/esm2015/lib/layout/mdl-layout-header.component.js
import { Component, ElementRef, HostBinding, HostListener, ViewEncapsulation, } from "@angular/core";
import { MdlLayoutMediatorService } from "./mdl-layout-mediator.service";
export class MdlLayoutHeaderComponent {
    constructor(elementRef, layoutMediatorService) {
        this.elementRef = elementRef;
        this.layoutMediatorService = layoutMediatorService;
        this.isSeamed = false;
        this.isLayoutHeader = true;
        this.isCompact = false;
        this.isAnimating = false;
        this.isRipple = true;
        this.el = elementRef.nativeElement;
    }
    get isWaterfall() {
        return this.mode === "waterfall";
    }
    get isCastingShadow() {
        return this.mode === "standard" || this.isCompact;
    }
    get isHeaderScroll() {
        return this.mode === "scroll";
    }
    onTransitionEnd() {
        this.isAnimating = false;
    }
    onClick() {
        if (this.isCompact) {
            this.isCompact = false;
            this.isAnimating = true;
        }
    }
    onTabMouseover(tab) {
        this.layoutMediatorService.tabMouseover(tab);
    }
    onTabMouseout(tab) {
        this.layoutMediatorService.tabMouseout(tab);
    }
    tabSelected(tab) {
        this.layoutMediatorService.tabSelected(tab);
    }
}
MdlLayoutHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-layout-header",
                template: `
    <ng-content></ng-content>
    <div
      *ngIf="tabs?.toArray()?.length > 0"
      class="mdl-layout__tab-bar-container"
    >
      <div class="mdl-layout__tab-bar is-casting-shadow">
        <div
          *ngFor="let tab of tabs.toArray()"
          class="mdl-layout__tab"
          [ngClass]="{ 'is-active': tab.isActive }"
          (mouseover)="onTabMouseover(tab)"
          (mouseout)="onTabMouseout(tab)"
        >
          <div
            *ngIf="tab.titleComponent"
            (click)="tabSelected(tab)"
            [mdl-ripple]="isRipple"
            [append-view-container-ref]="tab.titleComponent.vcRef"
          ></div>
          <a
            *ngIf="!tab.titleComponent"
            href="javascript:void(0)"
            (click)="tabSelected(tab)"
            class="mdl-layout__tab"
            [ngClass]="{ 'is-active': tab.isActive }"
            [mdl-ripple]="isRipple"
            >{{ tab.title }}</a
          >
        </div>
      </div>
    </div>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlLayoutHeaderComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: MdlLayoutMediatorService }
];
MdlLayoutHeaderComponent.propDecorators = {
    isSeamed: [{ type: HostBinding, args: ["class.mdl-layout__header--seamed",] }],
    isLayoutHeader: [{ type: HostBinding, args: ["class.mdl-layout__header",] }],
    isCompact: [{ type: HostBinding, args: ["class.is-compact",] }],
    isWaterfall: [{ type: HostBinding, args: ["class.mdl-layout__header--waterfall",] }],
    isCastingShadow: [{ type: HostBinding, args: ["class.is-casting-shadow",] }],
    isHeaderScroll: [{ type: HostBinding, args: ["class.mdl-layout__header--scroll",] }],
    onTransitionEnd: [{ type: HostListener, args: ["transitionend",] }],
    onClick: [{ type: HostListener, args: ["click",] }]
};
