// @angular-mdl/core/esm2015/lib/layout/mdl-layout-tab-panel.component.js
import { Component, ContentChild, HostBinding, Input, ViewEncapsulation, } from "@angular/core";
import { MdlTabPanelTitleComponent } from "../tabs/mdl-tab-panel-title.component";
export class MdlLayoutTabPanelComponent {
    constructor() {
        this.isLayoutTabPanel = true;
        this.isActive = false;
    }
}
MdlLayoutTabPanelComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-layout-tab-panel",
                template: `
    <ng-content
      *ngIf="titleComponent"
      select="mdl-tab-panel-content"
    ></ng-content>
    <ng-content *ngIf="!titleComponent"></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlLayoutTabPanelComponent.propDecorators = {
    titleComponent: [{ type: ContentChild, args: [MdlTabPanelTitleComponent, { static: true },] }],
    title: [{ type: Input, args: ["mdl-layout-tab-panel-title",] }],
    isLayoutTabPanel: [{ type: HostBinding, args: ["class.mdl-layout__tab-panel",] }],
    isActive: [{ type: HostBinding, args: ["class.is-active",] }]
};
