// @angular-mdl/core/esm2015/lib/chips/mdl-chip.component.js
import { Component, EventEmitter, HostBinding, Input, Output, ViewEncapsulation, } from "@angular/core";
export class MdlChipComponent {
    constructor() {
        // eslint-disable-next-line
        this.actionClick = new EventEmitter();
        this.isChip = true;
        this.isChipContact = false;
    }
    action() {
        this.actionClick.emit();
    }
}
MdlChipComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-chip",
                template: `
    <ng-content></ng-content>
    <span *ngIf="mdlLabel" class="mdl-chip__text">{{ mdlLabel }}</span>
    <button
      *ngIf="mdlActionIcon"
      (click)="action()"
      type="button"
      class="mdl-chip__action"
    >
      <mdl-icon>{{ mdlActionIcon }}</mdl-icon>
    </button>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlChipComponent.propDecorators = {
    mdlLabel: [{ type: Input, args: ['mdl-label',] }],
    mdlActionIcon: [{ type: Input, args: ['mdl-action-icon',] }],
    actionClick: [{ type: Output, args: ['action-click',] }],
    isChip: [{ type: HostBinding, args: ["class.mdl-chip",] }],
    isChipContact: [{ type: HostBinding, args: ["class.mdl-chip--contact",] }]
};
