// @angular-mdl/core/esm2015/lib/table/mdl-table.component.js
import { Component, EventEmitter, Input, Output, ViewEncapsulation, } from "@angular/core";
export class MdlDefaultTableModel {
    constructor(columns) {
        this.data = [];
        this.columns = columns;
    }
    addAll(data) {
        this.data.push(...data);
    }
}
const template = `
        <table class="mdl-data-table">
           <thead>
           <tr>
              <th *ngIf="selectable">
                 <mdl-checkbox mdl-ripple [ngModel]="isAllSelected()" (ngModelChange)="toogleAll()"></mdl-checkbox>
              </th>
              <th *ngFor="let column of model.columns"
                  [ngClass]="{'mdl-data-table__cell--non-numeric': !column.numeric}">
                 {{column.name}}
              </th>
           </tr>
           </thead>
           <tbody>
           <tr *ngFor="let data of model.data; let i = index" [ngClass]="{'is-selected': selectable && data.selected}">
              <td *ngIf="selectable">
                 <mdl-checkbox mdl-ripple
                      [(ngModel)]="data.selected"
                      (ngModelChange)="selectionChanged()"></mdl-checkbox>
              </td>
              <td *ngFor="let column of model.columns"
                  [ngClass]="{'mdl-data-table__cell--non-numeric': !column.numeric}"
                  [innerHTML]="data[column.key]">
              </td>
           </tr>
           </tbody>
        </table>
    `;
const styles = `
    :host{
      display:inline-block;
    }
    `;
export class MdlTableComponent {
    constructor() {
        this.selectable = false;
    }
    isAllSelected() {
        return false;
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    toogleAll() { }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    selectionChanged() { }
}
MdlTableComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-table",
                template,
                encapsulation: ViewEncapsulation.None,
                styles: [styles]
            },] }
];
MdlTableComponent.propDecorators = {
    model: [{ type: Input, args: ['table-model',] }]
};
export class MdlSelectableTableComponent extends MdlTableComponent {
    constructor() {
        super(...arguments);
        // eslint-disable-next-line
        this.selectionChange = new EventEmitter();
        this.selectable = true;
        this.allSelected = false;
    }
    isAllSelected() {
        return this.model.data.every((data) => data.selected);
    }
    toogleAll() {
        const selected = !this.isAllSelected();
        this.model.data.forEach((data) => (data.selected = selected));
        this.updateSelected();
    }
    selectionChanged() {
        this.updateSelected();
    }
    updateSelected() {
        this.selected = this.model.data.filter((data) => data.selected);
        this.selectionChange.emit({ value: this.selected });
    }
}
MdlSelectableTableComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-table-selectable",
                template,
                encapsulation: ViewEncapsulation.None,
                styles: [styles]
            },] }
];
MdlSelectableTableComponent.propDecorators = {
    model: [{ type: Input, args: ['table-model',] }],
    selected: [{ type: Input, args: ['table-model-selected',] }],
    selectionChange: [{ type: Output, args: ['table-model-selectionChanged',] }]
};
