// @angular-mdl/core/esm2015/lib/spinner/mdl-spinner.component.js
import { Component, HostBinding, Input, ViewEncapsulation, } from "@angular/core";
import { toBoolean } from "../common/boolean-property";
export class MdlSpinnerComponent {
    constructor() {
        this.isSpinner = true;
        this.isUpgraded = true;
        this.layers = [1, 2, 3, 4];
        this.activeIntern = false;
        this.singleColorIntern = false;
    }
    get active() {
        return this.activeIntern;
    }
    set active(value) {
        this.activeIntern = toBoolean(value);
    }
    get singleColor() {
        return this.singleColorIntern;
    }
    set singleColor(value) {
        this.singleColorIntern = toBoolean(value);
    }
}
MdlSpinnerComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-spinner",
                template: `
    <div
      *ngFor="let layer of layers"
      class="mdl-spinner__layer mdl-spinner__layer-{{ layer }}"
    >
      <div class="mdl-spinner__circle-clipper mdl-spinner__left">
        <div class="mdl-spinner__circle"></div>
      </div>
      <div class="mdl-spinner__gap-patch">
        <div class="mdl-spinner__circle"></div>
      </div>
      <div class="mdl-spinner__circle-clipper mdl-spinner__right">
        <div class="mdl-spinner__circle"></div>
      </div>
    </div>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlSpinnerComponent.propDecorators = {
    isSpinner: [{ type: HostBinding, args: ["class.mdl-spinner",] }],
    isUpgraded: [{ type: HostBinding, args: ["class.is-upgraded",] }],
    active: [{ type: HostBinding, args: ["class.is-active",] }, { type: Input }],
    singleColor: [{ type: HostBinding, args: ["class.mdl-spinner--single-color",] }, { type: Input, args: ["single-color",] }]
};
