// @angular-mdl/core/esm2015/lib/list/mdl-list.module.js
import { NgModule } from "@angular/core";
import { MdlListComponent, MdlListItemAvatarDirective, MdlListItemComponent, MdlListItemIconDirective, MdlListItemPrimaryContentComponent, MdlListItemSecondaryActionComponent, MdlListItemSecondaryContentComponent, MdlListItemSecondaryInfoComponent, MdlListItemSubTitleComponent, MdlListItemTextBodyComponent, } from "./mdl-list.component";
const MDL_LIST_DIRECTIVES = [
    MdlListComponent,
    MdlListItemComponent,
    MdlListItemPrimaryContentComponent,
    MdlListItemIconDirective,
    MdlListItemAvatarDirective,
    MdlListItemSecondaryContentComponent,
    MdlListItemSecondaryActionComponent,
    MdlListItemSubTitleComponent,
    MdlListItemSecondaryInfoComponent,
    MdlListItemTextBodyComponent,
];
export class MdlListModule {
    static forRoot() {
        return {
            ngModule: MdlListModule,
            providers: [],
        };
    }
}
MdlListModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                exports: MDL_LIST_DIRECTIVES,
                declarations: MDL_LIST_DIRECTIVES,
            },] }
];
export * from "./mdl-list.component";
