// @angular-mdl/core/esm2015/lib/list/mdl-list.component.js
import { Component, Directive, HostBinding, Input, Optional, ViewEncapsulation, } from "@angular/core";
import { MdlError, MdlStructureError } from "../common/mdl-error";
import { toNumber } from "../common/number.property";
export class MdlUnsupportedCountOfListItemLinesError extends MdlError {
    constructor(lines) {
        /* istanbul ignore next */
        super(`"${lines}" is not supported - max 3 lines please.`);
    }
}
export class MdlListComponent {
    constructor() {
        this.isMdlList = true;
    }
}
MdlListComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-list",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlListComponent.propDecorators = {
    isMdlList: [{ type: HostBinding, args: ["class.mdl-list",] }]
};
export class MdlListItemComponent {
    constructor() {
        this.isMdlListItem = true;
        this.linesIntern = 1;
    }
    get lines() {
        return this.linesIntern;
    }
    set lines(value) {
        this.linesIntern = toNumber(value);
    }
    get lines2() {
        return this.lines === 2;
    }
    get lines3() {
        return this.lines === 3;
    }
    ngOnChanges() {
        if (this.lines && this.lines > 3) {
            throw new MdlUnsupportedCountOfListItemLinesError(this.lines);
        }
    }
}
MdlListItemComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-list-item",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlListItemComponent.propDecorators = {
    isMdlListItem: [{ type: HostBinding, args: ["class.mdl-list__item",] }],
    lines: [{ type: Input }],
    lines2: [{ type: HostBinding, args: ["class.mdl-list__item--two-line",] }],
    lines3: [{ type: HostBinding, args: ["class.mdl-list__item--three-line",] }]
};
export class MdlListItemPrimaryContentComponent {
    constructor(mdlListItemComponent) {
        this.mdlListItemComponent = mdlListItemComponent;
        this.isPrimaryContent = true;
    }
    ngOnInit() {
        if (this.mdlListItemComponent === null) {
            throw new MdlStructureError("mdl-list-item-primary-content", "mdl-list-item");
        }
    }
}
MdlListItemPrimaryContentComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-list-item-primary-content",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlListItemPrimaryContentComponent.ctorParameters = () => [
    { type: MdlListItemComponent, decorators: [{ type: Optional }] }
];
MdlListItemPrimaryContentComponent.propDecorators = {
    isPrimaryContent: [{ type: HostBinding, args: ["class.mdl-list__item-primary-content",] }]
};
export class MdlListItemSecondaryContentComponent {
    constructor(mdlListItemComponent) {
        this.mdlListItemComponent = mdlListItemComponent;
        this.isSecondaryContent = true;
    }
    ngOnInit() {
        if (this.mdlListItemComponent === null) {
            throw new MdlStructureError("mdl-list-item-secondary-content", "mdl-list-item");
        }
    }
}
MdlListItemSecondaryContentComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-list-item-secondary-content",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlListItemSecondaryContentComponent.ctorParameters = () => [
    { type: MdlListItemComponent, decorators: [{ type: Optional }] }
];
MdlListItemSecondaryContentComponent.propDecorators = {
    isSecondaryContent: [{ type: HostBinding, args: ["class.mdl-list__item-secondary-content",] }]
};
export class MdlListItemSecondaryActionComponent {
    constructor(mdlListItemComponent) {
        this.mdlListItemComponent = mdlListItemComponent;
        this.isSecondaryAction = true;
    }
    ngOnInit() {
        if (this.mdlListItemComponent === null) {
            throw new MdlStructureError("mdl-list-item-secondary-action", "mdl-list-item");
        }
    }
}
MdlListItemSecondaryActionComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-list-item-secondary-action",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlListItemSecondaryActionComponent.ctorParameters = () => [
    { type: MdlListItemComponent, decorators: [{ type: Optional }] }
];
MdlListItemSecondaryActionComponent.propDecorators = {
    isSecondaryAction: [{ type: HostBinding, args: ["class.mdl-list__item-secondary-action",] }]
};
export class MdlListItemSubTitleComponent {
    constructor(mdlListItemComponent) {
        this.mdlListItemComponent = mdlListItemComponent;
        this.isSubTitle = true;
    }
    ngOnInit() {
        if (this.mdlListItemComponent === null) {
            throw new MdlStructureError("mdl-list-item-sub-title", "mdl-list-item-primary-content");
        }
    }
}
MdlListItemSubTitleComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-list-item-sub-title",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlListItemSubTitleComponent.ctorParameters = () => [
    { type: MdlListItemPrimaryContentComponent, decorators: [{ type: Optional }] }
];
MdlListItemSubTitleComponent.propDecorators = {
    isSubTitle: [{ type: HostBinding, args: ["class.mdl-list__item-sub-title",] }]
};
export class MdlListItemSecondaryInfoComponent {
    constructor(mdlListItemComponent) {
        this.mdlListItemComponent = mdlListItemComponent;
        this.isSecondaryInfo = true;
    }
    ngOnInit() {
        if (this.mdlListItemComponent === null) {
            throw new MdlStructureError("mdl-list-item-secondary-info", "mdl-list-item-secondary-content");
        }
    }
}
MdlListItemSecondaryInfoComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-list-item-secondary-info",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlListItemSecondaryInfoComponent.ctorParameters = () => [
    { type: MdlListItemSecondaryContentComponent, decorators: [{ type: Optional }] }
];
MdlListItemSecondaryInfoComponent.propDecorators = {
    isSecondaryInfo: [{ type: HostBinding, args: ["class.mdl-list__item-secondary-info",] }]
};
export class MdlListItemTextBodyComponent {
    constructor(mdlListItemComponent) {
        this.mdlListItemComponent = mdlListItemComponent;
        this.isTextBody = true;
    }
    ngOnInit() {
        if (this.mdlListItemComponent === null) {
            throw new MdlStructureError("mdl-list-item-text-body", "mdl-list-item");
        }
    }
}
MdlListItemTextBodyComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-list-item-text-body",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlListItemTextBodyComponent.ctorParameters = () => [
    { type: MdlListItemComponent, decorators: [{ type: Optional }] }
];
MdlListItemTextBodyComponent.propDecorators = {
    isTextBody: [{ type: HostBinding, args: ["class.mdl-list__item-text-body",] }]
};
export class MdlListItemIconDirective {
    constructor() {
        this.isItemIcon = true;
    }
}
MdlListItemIconDirective.decorators = [
    { type: Directive, args: [{
                // eslint-disable-next-line
                selector: 'mdl-icon[mdl-list-item-icon]'
            },] }
];
MdlListItemIconDirective.propDecorators = {
    isItemIcon: [{ type: HostBinding, args: ["class.mdl-list__item-icon",] }]
};
export class MdlListItemAvatarDirective {
    constructor() {
        this.isItemAvatar = true;
    }
}
MdlListItemAvatarDirective.decorators = [
    { type: Directive, args: [{
                // eslint-disable-next-line
                selector: 'mdl-icon[mdl-list-item-avatar]'
            },] }
];
MdlListItemAvatarDirective.propDecorators = {
    isItemAvatar: [{ type: HostBinding, args: ["class.mdl-list__item-avatar",] }]
};
