// @angular-mdl/core/esm2015/lib/common/animations.js
export class NativeWebAnimationPlayer {
    constructor(element, keyframes, duration, easing) {
        this.element = element;
        this.keyframes = keyframes;
        this.duration = duration;
        this.easing = easing;
        this.onDoneCallback = [];
    }
    onDone(fn) {
        this.onDoneCallback.push(fn);
    }
    play() {
        const animation = this.element.animate(this.keyframes, {
            duration: this.duration,
            easing: this.easing,
            fill: "forwards",
        });
        animation.addEventListener("finish", () => this.onDoneCallback.forEach((fn) => fn()));
    }
}
export class NoopAnimationPlayer {
    constructor() {
        this.onDoneCallback = [];
    }
    onDone(fn) {
        this.onDoneCallback.push(fn);
    }
    play() {
        this.onDoneCallback.forEach((fn) => fn());
    }
}
export class Animations {
}
export class NativeWebAnimations {
    animate(element, keyframes, duration, easing) {
        return new NativeWebAnimationPlayer(element, keyframes, duration, easing);
    }
}
export class NoopWebAnimations {
    animate() {
        return new NoopAnimationPlayer();
    }
}
