// @angular-mdl/core/esm2015/lib/badge/mdl-badge.directive.js
import { Directive, ElementRef, HostBinding, Input, Renderer2, } from "@angular/core";
const DATA_BADE_ATTR = "data-badge";
export class MdlBadgeDirective {
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.isBadge = true;
        this.el = elementRef.nativeElement;
    }
    ngOnChanges() {
        if (this.mdlBadgeContent === null ||
            typeof this.mdlBadgeContent === "undefined") {
            this.renderer.removeAttribute(this.el, DATA_BADE_ATTR);
            return;
        }
        this.renderer.setAttribute(this.el, DATA_BADE_ATTR, this.mdlBadgeContent);
    }
}
MdlBadgeDirective.decorators = [
    { type: Directive, args: [{
                // eslint-disable-next-line
                selector: '[mdl-badge]'
            },] }
];
MdlBadgeDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
MdlBadgeDirective.propDecorators = {
    mdlBadgeContent: [{ type: Input, args: ["mdl-badge",] }],
    isBadge: [{ type: HostBinding, args: ["class.mdl-badge",] }]
};
export class MdlBadgeOverlapDirective {
    constructor() {
        this.isOverlapping = true;
    }
}
MdlBadgeOverlapDirective.decorators = [
    { type: Directive, args: [{
                // eslint-disable-next-line
                selector: '[mdl-badge-overlap]'
            },] }
];
MdlBadgeOverlapDirective.propDecorators = {
    isOverlapping: [{ type: HostBinding, args: ["class.mdl-badge--overlap",] }]
};
export class MdlBadgeNoBackgroundDirective {
    constructor() {
        this.isNoBackground = true;
    }
}
MdlBadgeNoBackgroundDirective.decorators = [
    { type: Directive, args: [{
                // eslint-disable-next-line
                selector: '[mdl-badge-no-background]'
            },] }
];
MdlBadgeNoBackgroundDirective.propDecorators = {
    isNoBackground: [{ type: HostBinding, args: ["class.mdl-badge--no-background",] }]
};
