// @angular-mdl/core/esm2015/lib/tabs/mdl-tab-panel.component.js
import { Component, ContentChild, HostBinding, Input, ViewEncapsulation, } from "@angular/core";
import { MdlTabPanelTitleComponent } from "./mdl-tab-panel-title.component";
export class MdlTabPanelContentComponent {
}
MdlTabPanelContentComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-tab-panel-content",
                template: "<ng-content></ng-content>",
                encapsulation: ViewEncapsulation.None
            },] }
];
export class MdlTabPanelComponent {
    constructor() {
        this.isTabPanel = true;
        this.isActive = false;
    }
}
MdlTabPanelComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-tab-panel",
                template: `
    <ng-content
      *ngIf="titleComponent"
      select="mdl-tab-panel-content"
    ></ng-content>
    <ng-content *ngIf="!titleComponent"></ng-content>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlTabPanelComponent.propDecorators = {
    titleComponent: [{ type: ContentChild, args: [MdlTabPanelTitleComponent, { static: true },] }],
    title: [{ type: Input, args: ['mdl-tab-panel-title',] }],
    disabled: [{ type: Input, args: ['disabled',] }],
    isTabPanel: [{ type: HostBinding, args: ["class.mdl-tabs__panel",] }],
    isActive: [{ type: HostBinding, args: ["class.is-active",] }]
};
