// @angular-mdl/core/esm2015/lib/tooltip/mdl-tooltip.component.js
import { Component, ElementRef, HostBinding, Input, Renderer2, ViewEncapsulation, } from "@angular/core";
import { MdlTooltipPositionService } from "./mdl-tooltip-position.service";
const IS_ACTIVE = "is-active";
export class MdlSimpleTooltipComponent {
    constructor(elRef, renderer, mdlTooltipPositionService) {
        this.elRef = elRef;
        this.renderer = renderer;
        this.mdlTooltipPositionService = mdlTooltipPositionService;
        this.large = false;
        this.isTooltip = true;
        this.active = false;
        this.element = elRef.nativeElement;
    }
    get isBottom() {
        return this.position === "bottom";
    }
    get isRight() {
        return this.position === "right";
    }
    get isLeft() {
        return this.position === "left";
    }
    isTop() {
        return this.position === "top";
    }
    mouseLeave() {
        if (this.delayTimeout) {
            clearTimeout(this.delayTimeout);
        }
        this.setActive(false);
    }
    mouseEnter(event) {
        if (this.delay) {
            this.delayTimeout = setTimeout(() => {
                this.show(event.target);
            }, this.delay);
        }
        else {
            this.show(event.target);
        }
    }
    isActive() {
        return this.active;
    }
    show(element) {
        const props = element.getBoundingClientRect();
        const offsetWidth = this.element.offsetWidth;
        const offsetHeight = this.element.offsetHeight;
        const style = this.mdlTooltipPositionService.calcStyle(offsetWidth, offsetHeight, props, this.position);
        for (const key of Object.keys(style)) {
            this.renderer.setStyle(this.elRef.nativeElement, key, style[key]);
        }
        this.setActive(true);
    }
    setActive(active) {
        this.active = active;
        if (active) {
            this.renderer.addClass(this.elRef.nativeElement, IS_ACTIVE);
        }
        else {
            this.renderer.removeClass(this.elRef.nativeElement, IS_ACTIVE);
        }
    }
}
MdlSimpleTooltipComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-simple-tooltip",
                template: "<div>{{tooltipText}}</div>",
                providers: [MdlTooltipPositionService],
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlSimpleTooltipComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: MdlTooltipPositionService }
];
MdlSimpleTooltipComponent.propDecorators = {
    large: [{ type: HostBinding, args: ["class.mdl-tooltip--large",] }],
    position: [{ type: Input }],
    delay: [{ type: Input }],
    isTooltip: [{ type: HostBinding, args: ["class.mdl-tooltip",] }],
    isBottom: [{ type: HostBinding, args: ["class.mdl-tooltip--bottom",] }],
    isRight: [{ type: HostBinding, args: ["class.mdl-tooltip--right",] }],
    isLeft: [{ type: HostBinding, args: ["class.mdl-tooltip--left",] }],
    isTop: [{ type: HostBinding, args: ["class.mdl-tooltip--top",] }]
};
export class MdlTooltipComponent extends MdlSimpleTooltipComponent {
    constructor(elRef, renderer, mdlTooltipPositionService) {
        super(elRef, renderer, mdlTooltipPositionService);
    }
}
MdlTooltipComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-tooltip",
                template: "<div><ng-content></ng-content></div>",
                exportAs: "mdlTooltip",
                providers: [MdlTooltipPositionService],
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlTooltipComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: MdlTooltipPositionService }
];
