// @angular-mdl/core/esm2015/lib/tooltip/mdl-tooltip.directive.js
import { ComponentFactoryResolver, Directive, HostListener, Input, ViewContainerRef, } from "@angular/core";
import { MdlSimpleTooltipComponent, MdlTooltipComponent, } from "./mdl-tooltip.component";
export class AbstractMdlTooltipDirective {
    constructor(vcRef, large, componentFactoryResolver) {
        this.vcRef = vcRef;
        this.large = large;
        this.componentFactoryResolver = componentFactoryResolver;
    }
    onMouseEnter(event) {
        this.tooltipComponent.mouseEnter(event);
    }
    onMouseLeave() {
        this.tooltipComponent.mouseLeave();
    }
    ngOnInit() {
        // if the tooltip is not an instance of MdlTooltipComponent
        // we create a simpleTooltipComponent on the fly.
        if (!(this.tooltip instanceof MdlTooltipComponent)) {
            const cFactory = this.componentFactoryResolver.resolveComponentFactory(MdlSimpleTooltipComponent);
            const cRef = this.vcRef.createComponent(cFactory);
            this.tooltipComponent = cRef.instance;
            this.tooltipComponent.tooltipText = this.tooltip;
            this.configureTooltipComponent();
        }
        else {
            this.tooltipComponent = this.tooltip;
            this.configureTooltipComponent();
        }
    }
    ngOnChanges(changes) {
        if (changes.tooltip && !changes.tooltip.isFirstChange()) {
            if (!(this.tooltip instanceof MdlTooltipComponent)) {
                this.tooltipComponent.tooltipText = this.tooltip;
            }
        }
    }
    configureTooltipComponent() {
        this.tooltipComponent.large = this.large;
        this.tooltipComponent.position = this.position;
    }
}
AbstractMdlTooltipDirective.decorators = [
    { type: Directive }
];
AbstractMdlTooltipDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Boolean },
    { type: ComponentFactoryResolver }
];
AbstractMdlTooltipDirective.propDecorators = {
    onMouseEnter: [{ type: HostListener, args: ["touchend", ["$event"],] }, { type: HostListener, args: ["mouseenter", ["$event"],] }],
    onMouseLeave: [{ type: HostListener, args: ["window:touchstart",] }, { type: HostListener, args: ["mouseleave",] }]
};
export class MdlTooltipDirective extends AbstractMdlTooltipDirective {
    constructor(vcRef, componentFactoryResolver) {
        super(vcRef, false, componentFactoryResolver);
    }
}
MdlTooltipDirective.decorators = [
    { type: Directive, args: [{
                // eslint-disable-next-line
                selector: '[mdl-tooltip]'
            },] }
];
MdlTooltipDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: ComponentFactoryResolver }
];
MdlTooltipDirective.propDecorators = {
    tooltip: [{ type: Input, args: ["mdl-tooltip",] }],
    position: [{ type: Input, args: ["mdl-tooltip-position",] }]
};
export class MdlTooltipLargeDirective extends AbstractMdlTooltipDirective {
    constructor(vcRef, componentFactoryResolver) {
        super(vcRef, true, componentFactoryResolver);
    }
}
MdlTooltipLargeDirective.decorators = [
    { type: Directive, args: [{
                // eslint-disable-next-line
                selector: '[mdl-tooltip-large]'
            },] }
];
MdlTooltipLargeDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: ComponentFactoryResolver }
];
MdlTooltipLargeDirective.propDecorators = {
    tooltip: [{ type: Input, args: ["mdl-tooltip-large",] }],
    position: [{ type: Input, args: ["mdl-tooltip-position",] }]
};
