// @angular-mdl/core/esm2015/lib/switch/mdl-switch.component.js
import { Component, ElementRef, forwardRef, HostBinding, Renderer2, ViewEncapsulation, } from "@angular/core";
import { NG_VALUE_ACCESSOR } from "@angular/forms";
import { MdlCheckboxComponent } from "../checkbox/mdl-checkbox.component";
export class MdlSwitchComponent extends MdlCheckboxComponent {
    constructor(elementRef, renderer) {
        super(elementRef, renderer);
        this.isSwitch = true;
        this.isCheckbox = false;
    }
}
MdlSwitchComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-switch",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => MdlSwitchComponent),
                        multi: true,
                    },
                ],
                template: `
    <input
      type="checkbox"
      class="mdl-switch__input"
      (focus)="onFocus()"
      (blur)="onBlur()"
      [disabled]="disabled"
      [(ngModel)]="value"
    />
    <span class="mdl-switch__label"><ng-content></ng-content></span>
    <div class="mdl-switch__track"></div>
    <div class="mdl-switch__thumb">
      <span class="mdl-switch__focus-helper"></span>
    </div>
  `,
                encapsulation: ViewEncapsulation.None
            },] }
];
MdlSwitchComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
MdlSwitchComponent.propDecorators = {
    isSwitch: [{ type: HostBinding, args: ["class.mdl-switch",] }]
};
