// @angular-mdl/core/esm2015/lib/dialog-outlet/mdl-dialog-outlet.module.js
import { NgModule } from "@angular/core";
import { MdlDialogOutletComponent } from "./mdl-dialog-outlet.component";
import { MdlDialogOutletService } from "./mdl-dialog-outlet.service";
import { MdlBackdropOverlayComponent } from "./mdl-backdrop-overlay.component";
import { MdlDialogInnerOutletComponent } from "./mdl-dialog-inner-outlet.component";
export * from "./mdl-dialog-outlet.component";
export * from "./mdl-dialog-outlet.service";
export * from "./mdl-backdrop-overlay.component";
const PUBLIC_COMPONENTS = [MdlDialogInnerOutletComponent];
const PRIVATE_COMPONENTS = [
    MdlDialogOutletComponent,
    MdlBackdropOverlayComponent,
];
export class MdlDialogOutletModule {
    static forRoot() {
        return {
            ngModule: MdlDialogOutletModule,
            providers: [MdlDialogOutletService],
        };
    }
}
MdlDialogOutletModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                exports: PUBLIC_COMPONENTS,
                declarations: [...PUBLIC_COMPONENTS, ...PRIVATE_COMPONENTS],
                entryComponents: [MdlDialogOutletComponent, MdlBackdropOverlayComponent],
            },] }
];
