// @angular-mdl/core/esm2015/lib/dialog-outlet/mdl-backdrop-overlay.component.js
import { Component, EventEmitter, HostBinding, HostListener, NgZone, ViewEncapsulation, } from "@angular/core";
export class MdlBackdropOverlayComponent {
    constructor(ngZone) {
        this.ngZone = ngZone;
        this.zIndex = 0;
        this.isBackdrop = true;
        this.clickEmitter = new EventEmitter();
        this.visible = false;
    }
    get display() {
        return this.visible ? null : "none";
    }
    onBackdropClick(e) {
        // this event runs not in angular zone of the main app. make sure it runs in the main angular zone
        // and change detection works
        this.ngZone.run(() => {
            this.clickEmitter.emit();
        });
        e.stopPropagation();
    }
    hide() {
        this.visible = false;
    }
    showWithZIndex(zIndex) {
        this.zIndex = zIndex;
        this.visible = true;
    }
}
MdlBackdropOverlayComponent.decorators = [
    { type: Component, args: [{
                selector: "mdl-backdrop-overlay",
                template: ``,
                encapsulation: ViewEncapsulation.None,
                styles: [`
      .dialog-backdrop {
        position: fixed;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        background: rgba(0, 0, 0, 0.1);
      }
    `]
            },] }
];
MdlBackdropOverlayComponent.ctorParameters = () => [
    { type: NgZone }
];
MdlBackdropOverlayComponent.propDecorators = {
    zIndex: [{ type: HostBinding, args: ["style.zIndex",] }],
    isBackdrop: [{ type: HostBinding, args: ["class.dialog-backdrop",] }],
    display: [{ type: HostBinding, args: ["style.display",] }],
    onBackdropClick: [{ type: HostListener, args: ["click", ["$event"],] }]
};
